waku/waku_archive/driver/postgres_driver/postgres_driver.nim

```diff
@@ -587,6 +534,206 @@ proc sleep*(s: PostgresDriver, seconds: int):
+        onFatalError("error when creating a new partition from empty state: " & $error)
+    else:
+      let newestPartitionRes = self.partitionMngr.getNewestPartition()
+      if newestPartitionRes.isErr():
+        onFatalError("could not get newest partition: " & $newestPartitionRes.error)
+
+      let newestPartition = newestPartitionRes.get()
+      if newestPartition.containsMoment(now):
+        debug "creating a new partition for the future"
+        ## The current used partition is the last one that was created.
+        ## Thus, let's create another partition for the future.
+        (await self.addPartition(newestPartition.getLastMoment(), PartitionsRangeInterval)).isOkOr:
+          onFatalError("could not add the next partition for 'now': " & $error)
+      elif now >= newestPartition.getLastMoment():
+        debug "creating a new partition to contain current messages"
+        ## There is no partition to contain the current time.
+        ## This happens if the node has been stopped for quite a long time.
+        ## Then, let's create the needed partition to contain 'now'.
+        (await self.addPartition(now, PartitionsRangeInterval)).isOkOr:
+          onFatalError("could not add the next partition: " & $error)
+
+    await sleepAsync(DefaultDatabasePartitionCheckTimeInterval)
+
+proc startPartitionFactory*(self: PostgresDriver,
+                            onFatalError: OnFatalErrorHandler) {.async.} =
+
+  self.futLoopPartitionFactory = self.loopPartitionFactory(onFatalError)
+
+
+proc removeOldestPartition(self: PostgresDriver,
+                           forceRemoval: bool = false, ## To allow cleanup in tests
+                           ):
+                           Future[ArchiveDriverResult[void]] {.async.}  =
+  ## Indirectly called from the retention policy
+
+  let oldestPartition = self.partitionMngr.getOldestPartition().valueOr:
+    return err("could not remove oldest partition: " & $error)
+
+  if not forceRemoval:
+    let now = times.now().toTime().toUnix()
+    let currentPartitionRes = self.partitionMngr.getPartitionFromDateTime(now)
+    if currentPartitionRes.isOk():
+      ## The database contains a partition that would store current messages.
+
+      if currentPartitionRes.get() == oldestPartition:
+        debug "Skipping to remove the current partition"
+        return ok()
+
+  ## In the following lines is where the partition removal happens.
+  ## Detach and remove the partition concurrently to not block the parent table (messages)
+  let detachPartitionQuery =
+    "ALTER TABLE messages DETACH PARTITION " & oldestPartition.getName() & " CONCURRENTLY;"
+  debug "removeOldestPartition", query = detachPartitionQuery
+  (await self.performWriteQuery(detachPartitionQuery)).isOkOr:
+    return err(fmt"error in {detachPartitionQuery}: " & $error)
+
+  ## Clear all data from the partition and retrieve back space to the operating system
+  let truncateQuery = "TRUNCATE TABLE " & oldestPartition.getName()
```

Comment:
Why do we need to truncate date from the table before dropping it? Seems double work for me.